CSharp/Library/Microsoft.Bot.Connector.Shared/AttachmentEx.cs

```diff
@@ -0,0 +1,34 @@
+﻿namespace Microsoft.Bot.Connector
+{
+    using System;
+
+    public partial class Attachment : IEquatable<Attachment>
+    {
+        public bool Equals(Attachment other)
+        {
+            return other != null
+                && this.ContentType == other.ContentType
+                && this.ContentUrl == other.ContentUrl
+                && object.Equals(this.Content, other.Content)
+                && object.Equals(this.Name, other.Name)
+                && object.Equals(this.ThumbnailUrl, other.ThumbnailUrl);
+
+        }
+
+        public override bool Equals(object other)
+        {
+            return this.Equals(other as Attachment);
+        }
+
+        public override int GetHashCode()
+        {
+            var code = this.ContentType.GetHashCode()
+                ^ this.ContentUrl.GetHashCode()
+                ^ (this.Content == null ? 13 : this.Content.GetHashCode())
```

Comment:
@msft-shahins these fields could be null.  Prime numbers ensure if any combinatuon of two fields are null, the object will not be equal to a different instance having other two fields null. http://stackoverflow.com/questions/263400/what-is-the-best-algorithm-for-an-overridden-system-object-gethashcode